File: src/m/solve/loadresultfromdisk.m
function [variable fpos]=loadresultfromdisk(filename,step,name,varargin)
%LOADRESULTFROMDISK - load specific result of solution sequence from disk file "filename"
%
%   Usage:
%      variable=loadresultfromdisk(filename,step,name);

	%Open file
	fid=fopen(filename,'rb');
	if(fid==-1),
		error(['loadresultsfromdisk error message: could not open ',filename,' for binary reading']);
	end

	if nargin==4,
		%Put the pointer on the right position in the file:
		fpos=varargin{1};
		fseek(fid,fpos,'bof');
	end

	while 1,

		%read field
		fpos=ftell(fid);
		[length,count]=fread(fid,1,'int');

		if count==0,
			break;
			%we are done, break;
		else
			fieldname=fread(fid,length,'char');
			fieldname=fieldname(1:end-1)';
			fieldname=char(fieldname);
			rtime=fread(fid,1,'double');
			rstep=fread(fid,1,'int');

			%check on field: 
			if ((step==rstep) & (strcmpi(name,fieldname))),
				%ok, go read the result really: 
				type=fread(fid,1,'int');
				M=fread(fid,1,'int');
				if type==1,
					field=fread(fid,M,'double');
				elseif type==2,
					field=fread(fid,M,'char');
					field=char(field(1:end-1)');
				elseif type==3,
					N=fread(fid,1,'int');
					field=fread(fid,[N M],'double')';
				elseif type==4,
					N=fread(fid,1,'int');
					field=fread(fid,[N M],'int')';
				elseif type==5,
					N=fread(fid,1,'int');
					fieldr=fread(fid,[N M],'double')';
					fieldi=fread(fid,[N M],'double')';
					field=complex(fieldr,fieldi);
				else
					error(['cannot read data of type ' num2str(type) ]);
				end
				variable=field;
				break;
			else
				%just skim to next results.
				type=fread(fid,1,'int');
				M=fread(fid,1,'int');
				if type==1,
					fseek(fid,8*M,'cof');
				elseif type==2,
					fseek(fid,M,'cof');
				elseif type==3,
					N=fread(fid,1,'int');
					fseek(fid,M*N*8,'cof');
				elseif type==4,
					N=fread(fid,1,'int');
					fseek(fid,M*N*4,cof);
				else
					error(['cannot read data of type ' num2str(type) ]);
				end
			end
		end
	end
	fclose(fid);

File: src/m/solve/loadresultfromdisk.py
import struct
import numpy as np

def loadresultfromdisk(filename, step, name, *args):  # {{{
    """LOADRESULTFROMDISK - load specific result of solution sequence from disk
    file "filename"

    Usage:
        variable = loadresultsfromdisk(filename, step, name)

    TODO:
    - Test this module against output of src/m/solve/loadresultsfromdisk.m
    """

    # Open file
    try:
        fid = open(filename, 'rb')
    except IOError:
        raise IOError("loadresultsfromdisk error message: could not open {} for binary reading".format(filename))

    if len(args) == 4:
        # Put the pointer on the right position in the file
        fpos = args[0]
        fid.seek(fpos)

    while True:
        # read field
        fpos = tell(fid)
        length = struct.unpack('i', fid.read(struct.calcsize('i')))[0]

        fieldname = struct.unpack('{}s'.format(length), fid.read(length))[0][:-1]
        fieldname = fieldname.decode()  #strings are binaries when stored so need to be converted back
        rtime = struct.unpack('d', fid.read(struct.calcsize('d')))[0]
        rstep = struct.unpack('i', fid.read(struct.calcsize('i')))[0]

        # TODO: Check number of characters unpacked and break if need be (see
        #       src/m/solve/loadresultfromdisk.py)

        if (rstep == step) and (fieldname == name):
            # ok, go read the result really
            datatype = struct.unpack('i', fid.read(struct.calcsize('i')))[0]
            M = struct.unpack('i', fid.read(struct.calcsize('i')))[0]
            if datatype == 1:
                field = np.array(struct.unpack('{}d'.format(M), fid.read(M * struct.calcsize('d'))), dtype=float)
            elif datatype == 2:
                field = struct.unpack('{}s'.format(M), fid.read(M))[0][:-1]
                field = field.decode()
            elif datatype == 3:
                N = struct.unpack('i', fid.read(struct.calcsize('i')))[0]
                field = np.zeros(shape=(M, N), dtype=float)
                for i in range(M):
                    field[i, :] = struct.unpack('{}d'.format(N), fid.read(N * struct.calcsize('d')))
            elif datatype == 4:
                N = struct.unpack('i', fid.read(struct.calcsize('i')))[0]
                field = np.zeros(shape=(M, N), dtype=int)
                for i in range(M):
                    field[i, :] = struct.unpack('{}i'.format(N), fid.read(N * struct.calcsize('i')))
            elif datatype == 5:
                # TODO:
                # - Check that the following results in the same output as 
                # MATLAB
                #
                N = struct.unpack('i', fid.read(struct.calcsize('i')))[0]
                fieldr = np.zeros(shape=(M, N), dtype=float)
                fieldi = np.zeros(shape=(M, N), dtype=float)
                for i in range(M):
                    fieldr[i, :] = struct.unpack('{}d'.format(N), fid.read(N * struct.calcsize('d')))
                    fieldi[i, :] = struct.unpack('{}d'.format(N), fid.read(N * struct.calcsize('d')))
                field = np.vectorize(complex)(fieldr, fieldi)
                print(field)
            else:
                raise TypeError("cannot read data of type {}".format(datatype))

            variable = field
            break
        else:
            # just skim to next results
            datatype = struct.unpack('i', fid.read(struct.calcsize('i')))[0]
            M = struct.unpack('i', fid.read(struct.calcsize('i')))[0]
            if datatype == 1:
                fid.seek(M * 8, 1)
            elif datatype == 2:
                fid.seek(M, 1)
            elif datatype == 3:
                N = struct.unpack('i', fid.read(struct.calcsize('i')))[0]
                fid.seek(N * M * 8, 1)
            elif datatype == 4:
                N = struct.unpack('i', fid.read(struct.calcsize('i')))[0]
                fid.seek(N * M * 4, 1)
            else:
                raise TypeError("cannot read data of type {}".format(datatype))

    fid.close()

    return (variable, fpos)
# }}}
